fda/extrapolation.py

```diff
@@ -0,0 +1,332 @@
+            else:
+                shape = (fdata.nsamples, eval_points.shape[1])
+
+            res = np.full(shape, self.fill_value)
+
+        else:
+
+            idx_ext = _extrapolation_index(eval_points, fdata.domain_range)
+
+            if fdata.ndim_image == 1:
+                eval_points = eval_points.reshape(eval_points.shape[0],
+                                                  eval_points.shape[1])
+
+            res =  fdata(eval_points, derivative=derivative,
+                         extrapolation=ExtrapolationType.NONE, keepdims=keepdims)
+
+            if res.ndim == 3:
+                res[idx_ext, :] = self.fill_value
+            else:
+                res[idx_ext] = self.fill_value
+
+        return res
+
+
+def _parse_extrapolation(extrapolation, fdata=None):
+    """Parse the argument `extrapolation` in 'evaluate'.
+
+    Args:
+        extrapolation (:class:´Extrapolator´, str or Callable): Argument
+            extrapolation to be parsed.
+        fdata (:class:´FData´): Object with the default extrapolation.
+
+    Returns:
+        (:class:´Extrapolator´ or Callable): Extrapolation method.
+
+    """
+
+    # If extrapolation is None returns the object default
+    if extrapolation is None:
+
+        if fdata is not None:
+            return fdata.extrapolation
+        else:
+            return None
+    elif callable(extrapolation):
+        return extrapolation
+
+    # If not callable parses from string or ExtrapolationType enum
+    type = ExtrapolationType(extrapolation)
+
+    if type is ExtrapolationType.NONE:
+        return None
+    elif type is ExtrapolationType.EXCEPTION:
+        return ExceptionExtrapolation()
+    elif type is ExtrapolationType.PERIODIC:
+        return PeriodicExtrapolation()
+    elif type is ExtrapolationType.BOUNDS:
+        return BoundaryExtrapolation()
+    elif type is ExtrapolationType.NAN:
+        return FillExtrapolation(np.nan)
+    elif type is ExtrapolationType.ZEROS:
+        return FillExtrapolation(0)
+
+
+def _extrapolation_index(eval_points, domain_range):
+    """Checks the points that need to be extrapolated.
+
+    Args:
+        eval_points (numpy.ndarray): Array with shape `n_eval_points` x
+            `ndim_image` with the evaluation points, or shape ´nsamples´ x
+            `n_eval_points` x `ndim_image` with different evaluation points
+            for each sample.
+
+        domain_range (list of tuples): List with the bounds of each dimenson
+            of the domain range.
+
+    Returns:
+
+        (numpy.ndarray): Array with boolean index. The positions with True in
+            the index are outside the domain range.
+
+    """
+
+    # Case all samples evaluated with same times
+    if eval_points.ndim == 2:
+        index = np.full(eval_points.shape[0], False)
+
+    else: # Different times for each sample
+        index = np.full((eval_points.shape[0], eval_points.shape[1]) , False)
+
+    for i in range(len(domain_range)):
```

Comment:
Why not `index < domain_range[:, 0] | index > domain_range[:, 1]` instead of the loop?

Reply:
I'm not sure if this line could work. I tried other possibilities to make the vectorisation but didn't work, because eval points could be a 2d or 3d matrix.
Anyway, the domain dimension will be 1 or 2 in general , this loop won't affect the performance.